contracts/utils/Strings.sol

```diff
@@ -115,4 +121,89 @@ library Strings {
     function equal(string memory a, string memory b) internal pure returns (bool) {
         return bytes(a).length == bytes(b).length && keccak256(bytes(a)) == keccak256(bytes(b));
     }
+
+    /**
+     * @dev Parse a decimal string and returns the value as a `uint256`.
+     *
+     * This function will revert if:
+     * - the string contains any character that is not in [0-9].
+     * - the result does not fit in a `uint256`.
+     */
+    function toUint(string memory input) internal pure returns (uint256) {
+        bytes memory buffer = bytes(input);
+
+        uint256 result = 0;
+        uint256 bufferLength = buffer.length;
+        for (uint256 i = 0; i < bufferLength; ++i) {
+            result *= 10; // will revert if overflow
+            result += _parseChr(buffer[i], 10);
+        }
+        return result;
+    }
+
+    /**
+     * @dev Parse a decimal string and returns the value as a `int256`.
+     *
+     * This function will revert if:
+     * - the string contains any character (outside the prefix) that is not in [0-9].
+     * - the result does not fit in a `int256`.
+     */
+    function toInt(string memory input) internal pure returns (int256) {
+        bytes memory buffer = bytes(input);
+
+        // check presence of a negative sign.
+        bool isNegative = bytes1(buffer) == 0x2d;
+        int8 factor = isNegative ? int8(-1) : int8(1);
+
+        int256 result = 0;
+        uint256 bufferLength = buffer.length;
+        for (uint256 i = SafeCast.toUint(isNegative); i < bufferLength; ++i) {
+            result *= 10; // will revert if overflow
+            result += factor * int8(_parseChr(buffer[i], 10)); // parseChr is at most 9, it fits into an int8
+        }
+        return result;
+    }
+
+    /**
+     * @dev Parse a hexadecimal string (with or without "0x" prefix), and returns the value as a `uint256`.
+     *
+     * This function will revert if:
+     * - the string contains any character (outside the prefix) that is not in [0-9a-fA-F].
+     * - the result does not fit in a `uint256`.
+     */
+    function hexToUint(string memory input) internal pure returns (uint256) {
+        bytes memory buffer = bytes(input);
+
+        // skip 0x prefix if present. Length check doesn't appear to be critical
```

Comment:
second sentence feels out of place
```suggestion
        // skip 0x prefix if present
```

Reply:
Just for the record, this is what the comment was about:

The string length may be less than 2. String could be empty, of just `"1"`. In some cases, doing a `input[1]` or even a `input[0]` would revert (out of bound acces). Doing `bytes2(input)` does not revert if the string is to short.

So we check the length to verify that it ok to read the prefix ? It turn out no.

- If the string is empty, then regardless of the result of that lookup (that could read dirty bytes), the loop will not run (because length == 0), and the result will be 0 -> that is ok
- If the string has length 1 then we have two options
  - the check identifies the prefix
    - that means the string is `"0"`, and there is a dirty `x` after. 
    - In that case we have an offset of 2, and the length is 1, so the for loop does not run and the function returns 0 -> that is ok
  - the check does not find the prefix
    - the only "digit" is read by the loop, and the result should be just fine
- If the string has length >= 2, then the prefix lookup is in the bounds

That is the long explanation (I'm happy its visible in the PR 😃) to something that is not really trivial, and can be missed, but missing it is not a risk. 

We may get questions about it though ...